scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,328 @@
+
+    # ranges looks like np.array([[0, 1], [2, 3]]) for
+    # \int^0_1 \int_2^3 f(x, y) dydx
+    # f is assumed to be a function from a function from an array of shape
+    # (eval_points, dim_input) where dim_input is the dimension of the input
+    # and should return an array of shape (eval_points, dim_output) where dim_output
+    # is the dimension of the output, e.g. for f(x, y) = (x + y)^alpha where
+    # alpha is one of [.2, .5, .8] would have dim_input=2 and dim_output=3
+    # TODO: consider better name than ranges, e.g. look at nquad
+    def estimate(self, f, ranges):
+        # TODO: need to output some sort of error when the rule and f have inconsistent
+        # dimensions
+
+        range_start_points = ranges[:, 0]
+        range_end_points = ranges[:, 1]
+        range_lengths = range_end_points - range_start_points
+
+        # The underlying cubature rule is presumed to work on the interval (-1, 1)
+        # We need to apply a linear change of coordinates to map each interval (a, b)
+        # to (-1, 1).
+        # We also need to multiply the weights by a scale factor equal to the
+        # determinant of the Jacobian for this change.
+        weight_scale_factor = np.prod(range_lengths / 2)
+
+        nodes_scaled = (self.nodes + 1) * (range_lengths / 2) + range_start_points
+        error_nodes_scaled = (
+            (self.error_nodes + 1) * (range_lengths / 2) + range_start_points
+        )
+
+        weights_scaled = self.weights * weight_scale_factor
+        error_weights_scaled = self.error_weights * weight_scale_factor
+
+        # Need to resize the weights so they have the right shape
+        integral_estimate = np.sum(
+            weights_scaled[:, np.newaxis] * f(nodes_scaled),
+            axis=0
+        )
+        error_estimate = abs(np.sum(
+            error_weights_scaled[:, np.newaxis] * f(error_nodes_scaled),
+            axis=0
+        ))
+
+        # TODO: not yet worrying about rounding error
+
+        return integral_estimate, error_estimate
+
+
+# For now, assuming that the underlying quadrature rules all have linear error
+# estimators so that we can combine the error nodes and error weights
+class ProductRule(CubatureRuleLinearError):
+    def __init__(self, base_rules):
+        # TODO: using meshgrid like this probably isn't right
+        self.nodes = _cartesian_product(
+            [np.meshgrid(rule.nodes)[0] for rule in base_rules]
+        )
+        self.error_nodes = _cartesian_product(
+            [np.meshgrid(rule.error_nodes)[0] for rule in base_rules]
+        )
+
+        self.weights = np.prod(
+            _cartesian_product([rule.weights for rule in base_rules]),
+            axis=1
+        )
+        self.error_weights = np.prod(
+            _cartesian_product([rule.error_weights for rule in base_rules]),
+            axis=1
+        )
+
+
+# Maybe in the future it would be better if we could specify the degree and then it
+# would calculate the weights.
```

Comment:
Actually, since these are everywhere (quadpack, quad_vec, papers), it's probably fine to add them as a table + references. There was a minor typo in one of the weights though IIRC.